docs/wiki/EL/devp2p.md

```diff
@@ -1,7 +1,326 @@
+
+
+#### Message Framing
+
+- **Frames encapsulate encrypted messages** for efficient and secure communication.
+- **Multiplexing** allows multiple protocols to run over a single RLPx connection.
+
+##### Frame Structure
+
+| Field | Description |
+|-------|------------|
+| `header-ciphertext` | AES-encrypted **header** containing frame metadata. |
+| `header-mac` | **MAC** over the header for integrity verification. |
+| `frame-ciphertext` | AES-encrypted **message data**. |
+| `frame-mac` | **MAC** over the encrypted message data. |
+
+##### MAC Calculation
+
+- Uses **two keccak256 MAC states** (one for **ingress**, one for **egress**).
+- The MAC state is updated as frames are sent or received.
+- Ensures **message integrity** and prevents **tampering**.
+
+
+#### Capability Messaging
+
+- **Capabilities** define the supported protocols on a given connection.
+- **Multiplexing** enables concurrent usage of multiple capabilities.
+
+##### Message Structure
+
+| Field | Description |
+|-------|------------|
+| `msg-id` | Unique identifier for the message type. |
+| `msg-data` | **RLP-encoded** message payload. |
+| `frame-size` | **Compressed size** of `msg-data`. |
+
+
+#### P2P Capability Messages
+
+- The **"p2p" capability** is **mandatory** and used for initial negotiation.
+
+#### Core Messages
+
+| Message | ID | Function |
+|---------|----|----------|
+| `Hello` | `0x00` | Announces supported capabilities. |
+| `Disconnect` | `0x01` | Initiates a graceful disconnection. |
+| `Ping` | `0x02` | Checks if the peer is alive. |
+| `Pong` | `0x03` | Responds to a `Ping`. |
+
+#### Disconnect Reasons
+
+| Code | Reason |
+|------|--------|
+| `0x00` | Requested disconnect. |
+| `0x02` | Protocol violation. |
+| `0x03` | Useless peer. |
+| `0x05` | Already connected. |
+| `0x06` | Incompatible protocol version. |
+| `0x09` | Unexpected identity. |
+
+
+
+
+### 4. Application-Level Subprotocols  
+
+- **RLPx supports multiple application-level subprotocols** that enable specialized communication between Ethereum nodes.
+- These subprotocols are **built on top of the RLPx transport layer** and are used for  data exchange, state synchronization, and light client support.
+
+#### Common Ethereum Subprotocols  
+
+| **Subprotocol** | **Purpose** |
+|---------------|------------|
+| **Ethereum Wire Protocol (`eth`)** | Handles **blockchain data exchange**, including block propagation and transaction relaying. |
+| **Ethereum Snapshot Protocol (`snap`)** | Used for **state synchronization**, allowing nodes to download portions of the state trie. |
+| **Light Ethereum Subprotocol (`les`)** | Supports **light clients**, enabling them to request data from full nodes without storing the full state. |
+| **Portal Network (`portal`)** | A decentralized **state, block, and transaction retrieval network** for lightweight clients. |
+
```

Comment:
Sources and extra resources are missing. Is this information based on devp2p spec or other document? List any relevant materials, devcon talks, etc